extensions/data-workspace/src/services/workspaceService.ts

```diff
@@ -0,0 +1,65 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as dataworkspace from 'dataworkspace';
+import * as path from 'path';
+import { IWorkspaceService } from '../common/interfaces';
+import { ProjectProviderRegistry } from '../common/projectProviderRegistry';
+
+const WorkspaceConfigurationName = 'dataworkspace';
+const ProjectsConfigurationName = 'projects';
+
+export class WorkspaceService implements IWorkspaceService {
+	async getAllProjectTypes(): Promise<dataworkspace.IProjectType[]> {
+		await this.ensureProviderExtensionLoaded();
+		const projectTypes: dataworkspace.IProjectType[] = [];
+		ProjectProviderRegistry.providers.forEach(provider => {
+			projectTypes.push(...provider.supportedProjectTypes);
+		});
+		return projectTypes;
+	}
+
+	async getProjectsInWorkspace(): Promise<string[]> {
+		if (vscode.workspace.workspaceFile) {
+			const projects = <string[]>vscode.workspace.getConfiguration(WorkspaceConfigurationName).get(ProjectsConfigurationName);
+			return projects.map(project => path.isAbsolute(project) ? project : path.join(vscode.workspace.rootPath!, project));
```

Comment:
Should (can?) absolute pathing be forced as a convention for registered project providers?  I could see switching back and forth - or forgetting to in places - being a source of bugs.

Reply:
this area needs more attention when I introduce the add project to workspace in the future PR, I will make sure it works.